src/tool/xmeta/cpp/models/enum_model.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <string>
+
+#include "base_model.h"
+
+namespace xlang
+{
+    namespace xmeta
+    {
+        enum enum_type
+        {
```

Comment:
enums can also use bool as their underlying type,

Reply:
This seems wrong. What is the value of an enum backed by a bool?

Reply:
strong typing: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Ri-typed

Reply:
Depends on the language. ECMA 335 CLI spec says:

II.14.3 Enums 
The symbols of an enum are represented by an underlying integer type:  one of { bool, char, int8, unsigned int8, int16, unsigned int16, int32, unsigned int32, int64, unsigned int64, native int, unsigned native int }

However, ECMA 334 C# specification says:

9.3.10 Enumeration types An enumeration type is a distinct type with named constants. Every enumeration type has an underlying type, which shall be byte, sbyte, short, ushort, int, uint, long or ulong.

Personally, I don't see much value in allowing a enum to have a bool underlying base type. I could vaguely see a rationale for char, but then we'd have trouble projecting such an enum into C#.

I'd vote for restricting to the C# allowed underlying types.

Reply:
Like Chris was saying, bool-backed enums are incredibly valuable for improving usability, readability, and correctness. As an example from Win32: `auto event = CreateEvent(nullptr, FALSE, FALSE, "Name");`. Let's change this to an auto-reset event: `auto event = CreateEvent(nullptr, FALSE, TRUE, "Name");`.

See the bug?

I accidentally changed the initial state, not the reset behavior. With vague bool parameters, it's too easy to forget what those two bools mean, and too easy to accidentally change the wrong one. This interface is far more usable and robust: `auto event = CreateEvent(nullptr, reset_behavior::auto, initial_state::non_signalled, "name");`.